plugins/node/opentelemetry-plugin-express/test/express.test.ts

```diff
@@ -138,6 +172,124 @@ describe('Express Plugin', () => {
+          app.use(function customMiddleware(req, res, next) {
+            for (let i = 0; i < 1000000; i++) {
+              continue;
+            }
+            res.status(200).end('middleware');
+          });
+        }
+      });
+      const port = (server.address() as AddressInfo).port;
+      assert.strictEqual(memoryExporter.getFinishedSpans().length, 0);
+      await tracer.withSpan(rootSpan, async () => {
+        const response = await httpRequest.get(
+          `http://localhost:${port}/toto/tata`
+        );
+        assert.strictEqual(response, 'middleware');
+        rootSpan.end();
+        assert.strictEqual(finishListenerCount, 3);
+        assert.notStrictEqual(
+          memoryExporter
+            .getFinishedSpans()
+            .find(span => span.name.includes('customMiddleware')),
+          undefined,
+          'no customMiddleware span'
+        );
+      });
+      server.close();
+    });
+
+    it('supports async middlewares', async () => {
+      const rootSpan = tracer.startSpan('rootSpan');
+      let finishListenerCount: number | undefined;
+      const server = await serverWithMiddleware(tracer, rootSpan, app => {
+        app.use((req, res, next) => {
+          res.on('finish', () => {
+            finishListenerCount = res.listenerCount('finish');
+          });
+          next();
+        });
+        for (let index = 0; index < 15; index++) {
+          // eslint-disable-next-line prefer-arrow-callback
+          app.use(function customMiddleware(req, res, next) {
+            setTimeout(() => {
+              next();
+            }, 50);
+          });
+        }
+      });
+      const port = (server.address() as AddressInfo).port;
+      assert.strictEqual(memoryExporter.getFinishedSpans().length, 0);
+      await tracer.withSpan(rootSpan, async () => {
+        const response = await httpRequest.get(
+          `http://localhost:${port}/toto/tata`
+        );
+        assert.strictEqual(response, 'tata');
+        rootSpan.end();
+        assert.strictEqual(finishListenerCount, 2);
+        assert.notStrictEqual(
+          memoryExporter
+            .getFinishedSpans()
+            .find(span => span.name.includes('customMiddleware')),
+          undefined,
+          'no customMiddleware span'
+        );
+      });
+      server.close();
+    });
+
+    it('supports async middlewares directly responding', async () => {
+      const rootSpan = tracer.startSpan('rootSpan');
+      let finishListenerCount: number | undefined;
+      const server = await serverWithMiddleware(tracer, rootSpan, app => {
+        app.use((req, res, next) => {
+          res.on('finish', () => {
+            finishListenerCount = res.listenerCount('finish');
+          });
+          next();
+        });
+        for (let index = 0; index < 15; index++) {
+          // eslint-disable-next-line prefer-arrow-callback
+          app.use(function customMiddleware(req, res, next) {
```

Comment:
please create an arrow function out of the loop scope and then just use it here, instead of creating 15 new callbacks functions

Reply:
I’m not entirely sure how the refactored code should look like. Could you elaborate? The test intentionally creates 15 middlewares to validate we don’t see any warnings with > 11 listeners.

While the middleware function could be shared, it somehow feels less real world like as usually each middleware is a different function with different functionality.

Reply:
you declare a function out of the loop and then use it inside loop.

Reply:
I get that, but then we are calling effectively the same middleware multiple times, which isn’t really like something somebody would do in a real application.

Reply:
I agree with johannes here. I don't think we need to worry too much about micro-optimizing the test code.